models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/management/OllamaModelManager.java

```diff
@@ -0,0 +1,108 @@
+* limitations under the License.
+*/
+package org.springframework.ai.ollama.management;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.ai.ollama.api.OllamaApi;
+import org.springframework.ai.ollama.api.OllamaApi.DeleteModelRequest;
+import org.springframework.ai.ollama.api.OllamaApi.ListModelResponse;
+import org.springframework.ai.ollama.api.OllamaApi.PullModelRequest;
+import org.springframework.util.CollectionUtils;
+import reactor.util.retry.Retry;
+
+import java.time.Duration;
+
+/**
+ * Manage the lifecycle of models in Ollama.
+ *
+ * @author Christian Tzolov
+ * @author Thomas Vitale
+ * @since 1.0.0
+ */
+public class OllamaModelManager {
+
+	private final Logger logger = LoggerFactory.getLogger(OllamaModelManager.class);
+
+	private final OllamaApi ollamaApi;
+
+	private final ModelManagementOptions options;
+
+	public OllamaModelManager(OllamaApi ollamaApi) {
+		this(ollamaApi, ModelManagementOptions.defaults());
+	}
+
+	public OllamaModelManager(OllamaApi ollamaApi, ModelManagementOptions options) {
+		this.ollamaApi = ollamaApi;
+		this.options = options;
+	}
+
+	public boolean isModelAvailable(String modelName) {
+		ListModelResponse listModelResponse = ollamaApi.listModels();
+		if (!CollectionUtils.isEmpty(listModelResponse.models())) {
+			// Not an equality check to support the implicit ":latest" tag.
+			return listModelResponse.models().stream().anyMatch(m -> m.name().contains(modelName));
+		}
+		return false;
+	}
+
+	public void deleteModel(String modelName) {
+		logger.info("Start deletion of model: {}", modelName);
+		if (!isModelAvailable(modelName)) {
+			logger.info("Model {} not found", modelName);
+			return;
+		}
+		this.ollamaApi.deleteModel(new DeleteModelRequest(modelName));
+		logger.info("Completed deletion of model: {}", modelName);
+	}
+
+	public void pullModel(String modelName) {
+		pullModel(modelName, options.pullModelStrategy());
+	}
+
+	public void pullModel(String modelName, PullModelStrategy pullModelStrategy) {
+		if (PullModelStrategy.NEVER.equals(pullModelStrategy)) {
+			return;
+		}
+
+		if (PullModelStrategy.WHEN_MISSING.equals(pullModelStrategy)) {
+			if (isModelAvailable(modelName)) {
+				logger.debug("Model '{}' already available. Skipping pull operation.", modelName);
+				return;
+			}
+		}
+
+		// @formatter:off
+
+		logger.info("Start pulling model: {}", modelName);
+		this.ollamaApi.pullModel(new PullModelRequest(modelName))
+				.bufferUntilChanged(OllamaApi.ProgressResponse::status)
+				.doOnEach(signal -> {
+					var progressResponses = signal.get();
+					if (!CollectionUtils.isEmpty(progressResponses) && progressResponses.get(progressResponses.size() - 1) != null) {
+						logger.info("Pulling the '{}' model - Status: {}", modelName, progressResponses.get(progressResponses.size() - 1).status());
+					}
+				})
+				.takeUntil(progressResponses -> progressResponses.get(0) != null && progressResponses.get(0).status().equals("success"))
```

Comment:
Here we continuously print out the current status of the download, and the whole operation is configured with timeout and retry.